R/layer.R

```diff
@@ -475,19 +475,42 @@ check_subclass <- function(x, subclass,
                            env = parent.frame(),
                            call = caller_env()) {
   if (inherits(x, subclass)) {
-    x
-  } else if (is_scalar_character(x)) {
-    name <- paste0(subclass, camelize(x, first = TRUE))
-    obj <- find_global(name, env = env)
-
-    if (is.null(obj) || !inherits(obj, subclass)) {
-      cli::cli_abort("Can't find {argname} called {.val {x}}.", call = call)
-    } else {
-      obj
-    }
-  } else {
+    return(x)
+  }
+  if (!is_scalar_character(x)) {
     stop_input_type(x, as_cli("either a string or a {.cls {subclass}} object"))
   }
+
+  # Try getting class object directly
+  name <- paste0(subclass, camelize(x, first = TRUE))
+  obj <- find_global(name, env = env)
+  if (inherits(obj, subclass)) {
+    return(obj)
+  }
+
+  # Try retrieving class via constructors
+  name <- snakeize(name)
+  obj <- find_global(name, env = env, mode = "function")
+  if (is.function(obj)) {
+    obj <- obj()
```

Comment:
Maybe wrap in `try_fetch()` and rethrow the error with some additional context. We might end in situations where the constructor can't be called without arguments even though it generally should work